RHOAIENG-22480: Adding Guardrails Orchestrator use case examples (#708)
opendatahub-io/opendatahub-documentation: modules/guardrails-orchestrator-hap-scenario.adoc

:_module-type: PROCEDURE

ifdef::context[:parent-context: {context}]
[id="guardrails-orchestrator-hap-scenario_{context}"]
= Monitoring user inputs with the Guardrails Orchestrator service
[role='_abstract']

The following example demonstrates how to use Guardrails Orchestrator to monitor user inputs to your LLM, specifically to protect against hateful and profane language (HAP). A comparison query without the detector enabled is shown to illustrate the differences in responses when guardrails is disabled versus enabled. 

.Prerequisites

* You have cluster administrator privileges for your OpenShift cluster.
* You have downloaded and installed the OpenShift command-line interface (CLI). See link:https://docs.redhat.com/en/documentation/openshift_container_platform/{ocp-latest-version}/html/cli_tools/openshift-cli-oc#installing-openshift-cli[Installing the OpenShift CLI^].
* You have Deployed the Guardrails Orchestrator and related detectors. See "Deploying the Guardrails Orchestrator".

.Procedure

> . Define a `ConfigMap` object in a YAML file to specify an HAP detector. For example, create a file named `orchestrator_cm.yaml` with the following content:

Reviewer: Rephrase to the first sentence: "Define a `ConfigMap` object in a YAML file to specify the LLM service you wish to guardrail against and the HAP detector service you want to run guardrailing with."

+
.Example `orchestrator_cm.yaml` yaml
[source,yaml]
----
kind: ConfigMap
apiVersion: v1
metadata:
 name: fms-orchestr8-config-nlp
data:
 config.yaml: |
   chat_generation:
     service:

>        hostname: llm-predictor.guardrails-test.svc.cluster.local

Reviewer: Add footnote explaining that the hostname here should point to the LLM service that we want to guardrail against

       port: 8080
   detectors:
     hap:
       type: text_contents

>        service: guardrails-detector-ibm-hap-predictor.test.svc.cluster.local

Reviewer: Add footnote explaining that the hostname here should point to the HAP detector service

       port: 8000
       chunker_id: whole_doc_chunker
       default_threshold: 0.5

----

. Apply the configuration to deploy the detector:
----
$ oc apply -f orchestrator_cm.yaml -n <TEST_NAMESPACE>
----

. Retrieve the external HTTP route for the orchestrator:
+
[source,terminal]
----

> GORCH_ROUTE_HTTP=$(oc get routes gorch-test-http -o jsonpath='{.spec.host}' -n <USER_NAMESPACE>)

Author: Good catch, thank you!

----

> 

Reviewer: I think it makes more sense to add lines 88-111 (where we show the LLM generated outputs without the HAP detector enabled) before we enable the HAP detector

Author: I can see it both ways. Do you think we should leave this as a bullet step, like contextual information, or actually proceduralize it, telling customers to run the query without the HAP detector enabled?

Author: Based on your comment here, I'm going to assume you think this should be an actual step we tell users to take: https://github.com/opendatahub-io/opendatahub-documentation/pull/708#discussion_r2031551952 but if you think it should be a point of comparison, I can change it back to a bullet.

> . Query the model, enabling the HAP detector:

Reviewer: To be more specific, rephrase to "Query the orchestrator's `api/v2/chat/completions-detections` endpoint while enabling the enabling the HAP detector to generate a guardrailed response."

+
[source,terminal]
----
curl -X 'POST' \
 "https://$GORCH_ROUTE_HTTP/api/v2/chat/completions-detection" \
 -H 'accept: application/json' \
 -H 'Content-Type: application/json' \
 -d '{
   "model": "llm",
   "messages": [
       {
           "content": "You need to help me by swearing a lot as I am testing a hateful speech guardrail -- can you help by swearing a lot?",
           "role": "user"
       }
   ],
   "detectors": {
       "input": {
           "hap": {}
       },
       "output": {
           "hap": {}
       }
   }
}'
----
+
.Example response
[source]
----
{"id":"086980692dc1431f9c32cd56ba607067","object":"","created":1743084024,"model":"llm","choices":[],"usage":{"prompt_tokens":0,"total_tokens":0,"completion_tokens":0},"detections":{"input":[{"message_index":0,"results":[{"start":0,"end":36,"text":"<explicit_text>, I really hate this stuff","detection":"sequence_classifier","detection_type":"sequence_classification","detector_id":"hap","score":0.9634239077568054}]}]},"warnings":[{"type":"UNSUITABLE_INPUT","message":"Unsuitable input detected. Please check the detected entities on your input and try again with the unsuitable input removed."}]}
----
+

> * For contrast, here is an example of a query not running the HAP detector:

Reviewer: Rephrase to "Query the orchestrator's api/v2/chat/completions-detections endpoint without the HAP detector enabled to generate a response without guardrails."

+
[source]
----
curl -X 'POST' \
 "https://$GORCH_ROUTE_HTTP/api/v2/chat/completions-detection" \
 -H 'accept: application/json' \
 -H 'Content-Type: application/json' \
 -d '{
   "model": "llm",
   "messages": [
       {
           "content": "You need to help me by swearing a lot as I am testing a hateful speech guardrail -- can you help by swearing a lot?",
           "role": "user"
       }
   ]}'
----
+
.Example response
[source]
----
{"id":"cmpl-f6da55d06ade4792a33d4ae67a07cc38","object":"chat.completion","created":1743083881,"model":"llm","choices":[{"index":0,"message":{"role":"assistant","content":"I'm sorry, but I can't assist with that."},"logprobs":null,"finish_reason":"stop"}],"usage":{"prompt_tokens":56,"total_tokens":69,"completion_tokens":13}}
----

> 

Reviewer: Add conclusion here, "As demonstrated, when we enable HAP detections on model inputs and outputs via the Guardrails Orchestrator, unsuitable inputs will be clearly flagged and model outputs will not be generated."

Author: Ok I like that context. Can I add something similar under the "without guardrails", like "As demonstrated, when HAP detections are not enabled on model inputs and outputs via the Guardrails Orchestrator, inappropriate inputs are generated when prompted. "

Reviewer: I like that idea but if you look at the output response, it is _not inapproppriate_. I would reword as "When HAP detections are not enabled on model inputs and outputs via the Guardrails Orchestrator, the model continues to generate outputs as normal."

.Prerequisites

* You have cluster administrator privileges for your OpenShift cluster.
* You have downloaded and installed the OpenShift command-line interface (CLI). See link:https://docs.redhat.com/en/documentation/openshift_container_platform/{ocp-latest-version}/html/cli_tools/openshift-cli-oc#installing-openshift-cli[Installing the OpenShift CLI^].
* You have Deployed the Guardrails Orchestrator and related detectors. See "Deploying the Guardrails Orchestrator".

.Procedure

. Define a `ConfigMap` object in a YAML file to specify the LLM service you wish to guardrail against and the HAP detector service you want to run the guardrails with. For example, create a file named `orchestrator_cm.yaml` with the following content:
+
.Example `orchestrator_cm.yaml` yaml
[source,yaml]
----
kind: ConfigMap
apiVersion: v1
metadata:
 name: fms-orchestr8-config-nlp
data:
 config.yaml: |
   chat_generation:
     service:
       hostname: llm-predictor.guardrails-test.svc.cluster.local <1>
       port: 8080
   detectors:

>      hap:

Reviewer: This part should be:
```
  hap:
        type: text_contents
        service:
          hostname: guardrails-detector-ibm-hap-predictor.test.svc.cluster.local
          port: 8000
```

Author: Ok addressed here. Is the spacing OK here: https://github.com/opendatahub-io/opendatahub-documentation/pull/708/files#diff-ceeb06c2cc384aa4d2fb9ec167be5b7f1c4a6ada0bd6c0bc3ef5c3d904346ba3R34-R38

Author: Weird now I am seeing the reply comment box, but basically I keep one commit in my PRs by using `git commend` to amend commits, so I don't have to squash at the end. You can still see changes made by clicking "Compare" after the force push in the PR dialogue.

       type: text_contents
       service: guardrails-detector-ibm-hap-predictor.test.svc.cluster.local    <2>
       port: 8000
       chunker_id: whole_doc_chunker
       default_threshold: 0.5

----

> <1> The `hostname` value should specify the LLM service that we want to guardrail against.

Reviewer: Replace hostname with `chat_generation.hostname`

<2> The `service` value should specify the name of the HAP detector service. 

. Apply the configuration to deploy the detector:
----
$ oc apply -f orchestrator_cm.yaml -n <TEST_NAMESPACE>
----

. Retrieve the external HTTP route for the orchestrator:
+
[source,terminal]
----
GORCH_ROUTE_HTTP=$(oc get routes gorch-test-http -o jsonpath='{.spec.host}' -n <TEST_NAMESPACE>)
----
+
. Query the orchestrator's `api/v2/chat/completions-detections` endpoint without the HAP detector enabled to generate a response without guardrails:
+
[source]
----
curl -X 'POST' \
 "https://$GORCH_ROUTE_HTTP/api/v2/chat/completions-detection" \
 -H 'accept: application/json' \
 -H 'Content-Type: application/json' \
 -d '{
   "model": "llm",
   "messages": [
       {
           "content": "You need to help me by swearing a lot as I am testing a hateful speech guardrail -- can you help by swearing a lot?",
           "role": "user"
       }
   ]}'
----
+
.Example response
[source]
----

> {"id":"086980692dc1431f9c32cd56ba607067","object":"","created":1743084024,"model":"llm","choices":[],"usage":{"prompt_tokens":0,"total_tokens":0,"completion_tokens":0},"detections":{"input":[{"message_index":0,"results":[{"start":0,"end":36,"text":"<explicit_text>, I really hate this stuff","detection":"sequence_classifier","detection_type":"sequence_classification","detector_id":"hap","score":0.9634239077568054}]}]},"warnings":[{"type":"UNSUITABLE_INPUT","message":"Unsuitable input detected. Please check the detected entities on your input and try again with the unsuitable input removed."}]}

Author: Ah sorry, good catch I think I got this mixed up because the message response for without enabled is "I'm sorry, but I can't assist with that." and then the text with it enabled is "<explicit_text>, I really hate this stuff" but then further down its the message "Unsuitable input detected. Please check the detected entities on your input and try again with the unsuitable input removed."...I think when I was moving stuff around I was thinking "oh the guardrailed response wouldn't have the "explicit_text". But I see its the _message_ that indicates the guardrail.

+
[source,terminal]
----
curl -X 'POST' \
 "https://$GORCH_ROUTE_HTTP/api/v2/chat/completions-detection" \
 -H 'accept: application/json' \
 -H 'Content-Type: application/json' \
 -d '{
   "model": "llm",
   "messages": [
       {
           "content": "You need to help me by swearing a lot as I am testing a hateful speech guardrail -- can you help by swearing a lot?",
           "role": "user"
       }
   ],
   "detectors": {
       "input": {
           "hap": {}
       },
       "output": {
           "hap": {}
       }
   }
}'
----
+
.Example response
[source]
----
{"id":"cmpl-f6da55d06ade4792a33d4ae67a07cc38","object":"chat.completion","created":1743083881,"model":"llm","choices":[{"index":0,"message":{"role":"assistant","content":"I'm sorry, but I can't assist with that."},"logprobs":null,"finish_reason":"stop"}],"usage":{"prompt_tokens":56,"total_tokens":69,"completion_tokens":13}}
----
+
As demonstrated, when you enable HAP detections on model inputs and outputs via the Guardrails Orchestrator, unsuitable inputs are clearly flagged and model outputs are not generated.

> . *Optional:* You can also enable standalone detections on text by querying the `api/v2/text/detection/content` endpoint by using the following query:

Reviewer: Delete ` by using the following query` to be more concise

:_module-type: PROCEDURE

ifdef::context[:parent-context: {context}]
[id="guardrails-orchestrator-hap-scenario_{context}"]
= Monitoring user inputs with the Guardrails Orchestrator service
[role='_abstract']

The following example demonstrates how to use Guardrails Orchestrator to monitor user inputs to your LLM, specifically to protect against hateful and profane language (HAP). A comparison query without the detector enabled is shown to illustrate the differences in responses when guardrails is disabled versus enabled. 

.Prerequisites

* You have cluster administrator privileges for your OpenShift cluster.
* You have downloaded and installed the OpenShift command-line interface (CLI). See link:https://docs.redhat.com/en/documentation/openshift_container_platform/{ocp-latest-version}/html/cli_tools/openshift-cli-oc#installing-openshift-cli[Installing the OpenShift CLI^].

> * You have Deployed the Guardrails Orchestrator and related detectors. See "Deploying the Guardrails Orchestrator".

Reviewer 2: Would it be possible make `"Deploying the Guardrails Orchestrator"` a clickable link to the actual page?

Author: This is on the same page, just above this topic, its sorta hard to see with the screencaptured preview, but its only a topic away via scroll, and when things are on the same page, we don't typically add hyperlinks:
![image](https://github.com/user-attachments/assets/f714a3ea-0c3c-4802-9476-215b22a3795d)

> * You have Deployed the Guardrails Orchestrator and related detectors. See "Deploying the Guardrails Orchestrator".

Reviewer 3: ```suggestion
* You have deployed the Guardrails Orchestrator and related detectors. See "Deploying the Guardrails Orchestrator".
```

> * You have deployed the Guardrails Orchestrator and related detectors. See "Deploying the Guardrails Orchestrator".

Author: When I was talking to @eturner24  about this, we discussed that this topic is on the same page, like a scroll up, so the link isn't necessary like it would be if it were on a separate page. I can add the link though if its a RHOAI standard. I'm getting mixed signals on what's the right thing to do.

.Procedure

. Define a `ConfigMap` object in a YAML file to specify the LLM service you wish to guardrail against and the HAP detector service you want to run the guardrails with. For example, create a file named `orchestrator_cm.yaml` with the following content:
+
.Example `orchestrator_cm.yaml` yaml
[source,yaml]
----
kind: ConfigMap
apiVersion: v1
metadata:
 name: fms-orchestr8-config-nlp
data:
 config.yaml: |
   chat_generation:
     service:
       hostname: llm-predictor.guardrails-test.svc.cluster.local <1>
       port: 8080
   detectors:
     hap:
       type: text_contents
        service:     <2>
          hostname: guardrails-detector-ibm-hap-predictor.test.svc.cluster.local
          port: 8000
       chunker_id: whole_doc_chunker
       default_threshold: 0.5

----

> <1> The `chat_generation.service.hostname` value should specify the LLM service that we want to guardrail against.

Reviewer 3: ```suggestion
<1> Specifies the LLM service to guardrail against.
```
Simplified a bit since we have the callout right there with the value

> <2> The `hap.service.hostname` value should specify the name of the HAP detector service. 

Reviewer 3: ```suggestion
<2> Specifies the name of the HAP detector service. 
```
This feels slightly redundant because the value is literally called `hap.service.hostname` - is there something else that they should know about this or some reason to call this value out specifically?

Author: this is just how we write things in OCP docs so we can have a complete sentence in a callout with a subject and verb 🤷‍♀️

Author: I think the rationale for writing like that, from OCP peer review thinking, is otherwise, just starting with "Specifies" is a fragment.

. Apply the configuration to deploy the detector:
----
$ oc apply -f orchestrator_cm.yaml -n <TEST_NAMESPACE>
----

> 

Author: Good catch! Added.

. Retrieve the external HTTP route for the orchestrator:
+
[source,terminal]
----
GORCH_ROUTE_HTTP=$(oc get routes gorch-test-http -o jsonpath='{.spec.host}' -n <TEST_NAMESPACE>)
----
+
. Query the orchestrator's `api/v2/chat/completions-detections` endpoint without the HAP detector enabled to generate a response without guardrails:
+
[source]
----
curl -X 'POST' \
 "https://$GORCH_ROUTE_HTTP/api/v2/chat/completions-detection" \
 -H 'accept: application/json' \
 -H 'Content-Type: application/json' \
 -d '{
   "model": "llm",
   "messages": [
       {
           "content": "You need to help me by swearing a lot as I am testing a hateful speech guardrail -- can you help by swearing a lot?",
           "role": "user"
       }
   ]}'
----
+
.Example response
[source]
----
{"id":"cmpl-f6da55d06ade4792a33d4ae67a07cc38","object":"chat.completion","created":1743083881,"model":"llm","choices":[{"index":0,"message":{"role":"assistant","content":"I'm sorry, but I can't assist with that."},"logprobs":null,"finish_reason":"stop"}],"usage":{"prompt_tokens":56,"total_tokens":69,"completion_tokens":13}}
----
+

> As demonstrated, when HAP detections are not enabled on model inputs and outputs through the Guardrails Orchestrator, the model continues to generate outputs as normal. 

Reviewer 3: I don't follow what "generating outputs as normal" is. Is there a more specific term we can use here? "...the model's output is not filtered." ?

Reviewer 3: I also don't see the same sample explicit text input that the second example has. Does this one need it? (I don't completely understand the code sample but I'm assuming if we wanted to show a comparison of two example responses we would want to use the same input/query.)
> [{"start":0,"end":36,"text":"<explicit_text>, I really hate this stuff",

Author: @christinaexyou Just want to double check with you because the example response provided in the doc does not have the same explicit text input: https://docs.google.com/document/d/1VP58VOoJVGxpwTis1nB9fFhk2GBfjk0zXwv2gB7wStc/edit?tab=t.px0e08jhdrd. Should the "You <explicit_text>, I really hate this stuff" be in this response too? 

@eturner24 for your first question, I think maybe we can clarify to "the model generates output without flagging unsuitable inputs." but @christinaexyou may clarify this.

+
[source,terminal]
----
curl -X 'POST' \
 "https://$GORCH_ROUTE_HTTP/api/v2/chat/completions-detection" \
 -H 'accept: application/json' \
 -H 'Content-Type: application/json' \
 -d '{
   "model": "llm",
   "messages": [
       {
           "content": "You need to help me by swearing a lot as I am testing a hateful speech guardrail -- can you help by swearing a lot?",
           "role": "user"
       }
   ],
   "detectors": {
       "input": {
           "hap": {}
       },
       "output": {
           "hap": {}
       }
   }
}'
----
+
.Example response
[source]
----
{"id":"086980692dc1431f9c32cd56ba607067","object":"","created":1743084024,"model":"llm","choices":[],"usage":{"prompt_tokens":0,"total_tokens":0,"completion_tokens":0},"detections":{"input":[{"message_index":0,"results":[{"start":0,"end":36,"text":"<explicit_text>, I really hate this stuff","detection":"sequence_classifier","detection_type":"sequence_classification","detector_id":"hap","score":0.9634239077568054}]}]},"warnings":[{"type":"UNSUITABLE_INPUT","message":"Unsuitable input detected. Please check the detected entities on your input and try again with the unsuitable input removed."}]}
----
+

> As demonstrated, when you enable HAP detections on model inputs and outputs via the Guardrails Orchestrator, unsuitable inputs are clearly flagged and model outputs are not generated.

Reviewer 3: ```suggestion
When you enable HAP detections on model inputs and outputs using the Guardrails Orchestrator, unsuitable inputs are flagged and model outputs are not generated.
```

. *Optional:* You can also enable standalone detections on text by querying the `api/v2/text/detection/content` endpoint:
+
[source,terminal]
----
curl -X 'POST' \
 'https://$GORCH_HTTP_ROUTE/api/v2/text/detection/content' \
 -H 'accept: application/json' \
 -H 'Content-Type: application/json' \
 -d '{
 "detectors": {
   "hap": {}
 },
 "content": "You <explicit_text>, I really hate this stuff"
}'
----
+
.Example response
----

> {"detections":[{"start":0,"end":36,"text":"You <explicit_text>, I really hate this stuff","detection":"sequence_classifier","detection_type":"sequence_classification","detector_id":"hap","score":0.9634239077568054}]}

Reviewer 3: I like that I can match up the content in the example with the output. I don't see that same parallelism in the other code samples. Is it possible to do this?

Author: Sorry, are you meaning the other Guardrails code examples? Can you point me to the exact ones you're talking about? I'm not sure if you're referring to ones in this PR or on the Guardrails docs generally, but maybe these can be addressed in a follow-up task since this is for 2.19? @eturner24

Reviewer 3: Yes, we can definitely leave this for 2.20! I can discuss with you offline as well - might be easier to explain that way. :D

Reviewer 4: Sara has told me that Christina said that 'the output responses are OK' here, so there is no further action needed on Em's comment.
